Image Upload Fix

Allcharles opened the pull request:
# Image Upload Fix

Enable the ability for users to upload images for projects, regions, and sites

## Changes

- Rename `image` attribute to `images` for models (`image` was a the converted version of `imageUrls` with accurate url paths, and default values)
- Used `image` attribute for model property when uploading images
- Updated services to handle image uploading using multipart requests

## Issues

Fixes #608 

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

Allcharles committed:
Working image uploads

Allcharles committed:
Change model kinds to reflect baw-server

Allcharles committed:
Simplify BawPersistAttributeOptions and update comments

Allcharles committed:
Fix broken tests

Allcharles committed:
Cleanup baw api services

Allcharles committed:
Add TODO

Allcharles committed:
Final cleanup

atruskie commented on src/app/components/shared/formly/image-input.component.ts:
```diff
@@ -24,11 +24,20 @@ import { asFormControl } from "./helper";
+    if (imageFiles instanceof FileList && imageFiles.length > 0) {
+      this.formControl.setValue(imageFiles.item(0));
+    }
+  }
```
i wonder if this should throw if more than one file is present

if file `input` does support multi-select  then picking the first file would be really mis-representing the UI's intent

Allcharles replied:
It should not be possible to select multiple files, file input is set to only accept a single file

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -62,32 +74,36 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+        return true;
       }
-    });
+    }
```
personally i'd express this with `some`

```
this.getModelAttributes({ ...opts, formData: true })).values().some(isInstantiated)
```

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -62,32 +74,36 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+       */
+      const snakeCaseAttr = snakeCase(attr);
+      output.append(`${this.kind}[${snakeCaseAttr}]`, data[attr]);
+    }
```
Something is off with the naming here. 

Technically, every field (except notes - and even that is solvable with effort) could be sent via form data. I'd expect `toFormData` to work exactly the same as `toJson`, it's just that the data would be serialised differently.

I think we want to de-generalize this a bit. 

- if we're only using `toFormData` for image uploads then it should be a mixin, not defined on `AbstractModel`
- i think we should be a little more specific in naming with intent
  - method name could be something like `getImageAsFormData`
  - `getModelData`'s option could be `images` over `formData`

Needs some more thought, but unless there are cases I haven't considered, emphasis on being more specific and less general

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -106,12 +122,59 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+
+  /**
+   * Converts the model and a list of keys, into an object containing each key
+   * value pair
```
A better description might be:

```
Converts the model and its values into a simpler object representation which is suitable for serialization. Only the `keys` specified are included in the result.
```

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -106,12 +122,59 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+   * Converts the model and a list of keys, into an object containing each key
+   * value pair
+   *
+   * @param keys List of attributes to retrieve
```
```suggestion
   * @param keys List of attributes to extract
```

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -106,12 +122,59 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+   *
+   * @param keys List of attributes to retrieve
+   */
+  private toObject(keys: string[]): Partial<this> {
```
Hmm: the return type `Partial<this>` is not really true is it? Several value types are stringified in the result object, which notionally makes it a different type

Allcharles replied:
Technically it still is. The values are the same format as when the data came in which is defined in the Interface for the model. Seeing as this class extends its interface, a `Partial<this>` is correct. Ie. `IProject` defines `siteIds` as a `Id[] | Set<Id>`. In `toObject` we just convert the value from `Set<Id>` back to `Id[]`.

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -106,12 +122,59 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+   * not the injector
+   */
+  private getModelAttributes(opts?: {
+    create?: boolean;
+    update?: boolean;
+    formData?: boolean;
```
See previous comment, but the `formData` is really not expressing the intent well. You want to extract image related fields only, so you can send those via form data.

```suggestion
    images?: boolean;
```

atruskie replied:
personally i think these `opts` could just be parameters... there are no additional arguments needed to justify an options object

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -106,12 +122,59 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+        : "jsonAttributes";
+      return this[AbstractModel.keys.create[attribute]] ?? [];
+    } else if (opts?.update) {
+      const attribute: keyof typeof AbstractModel.keys.create = opts.formData
```
should this not be `.update` ?

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -106,12 +122,59 @@ export abstract class AbstractModelWithoutId<Model = Record<string, any>> {
+    if (opts?.create) {
+      const attribute: keyof typeof AbstractModel.keys.create = opts.formData
+        ? "formDataAttributes"
+        : "jsonAttributes";
```
this seems wrong... again, almost all attributes can be encoded in either json or form data

atruskie commented on src/app/models/AttributeDecorators.ts:
```diff
@@ -4,37 +4,55 @@ import { Id, Ids, ImageSizes, ImageUrl } from "@interfaces/apiInterfaces";
+  model[createKeys.jsonAttributes] ??= [];
+  model[updateKeys.jsonAttributes] ??= [];
+  model[createKeys.formDataAttributes] ??= [];
+  model[updateKeys.formDataAttributes] ??= [];
```
see main review comment: but this feels untenable

atruskie commented on src/app/models/AttributeDecorators.ts:
```diff
@@ -236,18 +254,27 @@ function createDecorator<Model>(
+      /** Include attribute in update requests for the model */
+      update?: boolean;
+      /** Serve this value in a form data request, instead of the usual json */
+      formData?: boolean;
```
again most values could be submitted to API as form data. 

This parameter would be better named `formDataOnly` (or see the `supprortedFormats`  suggestion from main review).

atruskie commented on src/app/models/Project.ts:
```diff
@@ -44,19 +45,21 @@ export interface IProject extends HasAllUsers, HasDescription {
   })
-  public readonly image: ImageUrl[];
+  public readonly images: ImageUrl[];
+  @bawPersistAttr({ create: true, update: true, formData: true })
```
a great example where arguments with default values would make a nicer api over an opts object

atruskie commented on src/app/services/baw-api/baw-api.service.spec.ts:
```diff
@@ -46,13 +47,14 @@ class IMockModel {
 }
 
 class MockModel extends AbstractModel implements IMockModel {
+  public kind = "MockModel";
```
i think camel case is incorrect, but if you convince me it's not then this field is inconsistent with the others

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -208,23 +217,58 @@ export class BawApiService<Model extends AbstractModel> {
+    updatePath: (model: Model) => string,
+    body: AbstractModel
+  ) {
+    const formData = body.toFormData({ create: true });
```
yeah this is where this naming issue really comes into it.

This method is really better named as `apiUpdateForFormDataOnlyAttributes`  since that is how it is used.

And `body.toFormData` is better named... well i don't have a better name, but it's intent is 'get only the fields that can't be updated as JSON and format a formData response`

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -208,23 +217,58 @@ export class BawApiService<Model extends AbstractModel> {
+    } else {
+      return this.httpPatch(path, jsonData ?? body).pipe(
+        map(this.handleSingleResponse)
+      );
```
this can probably be flattened into one chain

Allcharles replied:
Tried to, but I'm unsure of a good method to maintain type support. Potentially a const array, or maybe a function

Allcharles replied:
Nvm, figured out how

atruskie reviewed (changes requested):
As mentioned in chat: I think `ImageUrls` values should be normalised and updated and I don't think we need to alias the field to something else like `Images` (especially since the field is read-only).

I'm not fond of the naming in some of the `AbstractModel` methods. I think the correct naming depends on use cases which may not be visible in this PR, so open to discussion. However, you should split intent (get these things) out from serialization format (json/formdata) and not mix them.

An alternate design might actually be to store more metadata about attributes - rather than just the name.  i.e. `BawPersistAttr` would (rather than inserting a string into a number of different arrays) insert an object of type 

```typescript
{
  name: string,
  create: boolean,
  update: boolean,
  supportedFormats: Array<"json"|"formData">
} 
```

This second proposal would probably make it easier to get attributes for these different cases too.  `getFormDataCreateAttributes = (x) => x.create && x.supportedFormats.contains("formData") ; this.attributes.map(getFormDataCreateAttributes )`


Why did all the model `kind` fields get converted to _snake case_? The kind field is a type script only concept, and thus a client only concept, and should be in _camel case_ along with everything else.


So I think I want to see this flow readily visible in the services layer:

1. Get json serializable model
2. update/create
3. see if there are any formData only supported dirty attributes on the model
4. if yes, get the form data payload for form-data-only attributes
5. issue a second request, which will always be an update request, and which also will be multipart

This model should remove the need for `apiCreateMultipart` - `apiCreate` will just handle it when needed.

Allcharles committed:
Removed images property

Allcharles committed:
Reimplemented abstract model attributes

Allcharles committed:
Cleanup some review comments

- Log error when file input returns too many files
- Added tests for hasFormData
- Cleanup AbstractModel
- Simplify BawApiService

Allcharles committed:
Cleanup naming schemes

Allcharles committed:
Fixed broken tests

Allcharles committed:
Renamed whitelist to allow list

Allcharles committed:
Update libssl1.1 dependency
